File: server/shared/typescript/location/src/compose-location-hash.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import { sha256Hex } from "./internal/sha256-hex.js";
import type { AdminLocation } from "./value-objects/admin-location.js";
import type { Coordinates } from "./value-objects/coordinates.js";
import type { StreetAddress } from "./value-objects/street-address.js";

/**
 * Composes a single hash identifier from up to three location
 * components. Returns `undefined` when ALL three inputs are undefined /
 * null (location is absent — not an error). Otherwise returns
 * `"v1." + sha256(c.hashId | s.hashId | a.hashId)`; missing slots
 * contribute `""` (positional, never collapsed). Inner component
 * `"v1."` prefixes ARE included in the outer hash input.
 *
 * Returns `string | undefined` rather than `D2Result<string>` —
 * documented §17 carve-out: the operation cannot fail (inputs are
 * already-validated value objects or null/undefined), and the all-null
 * case is a legitimate non-error state.
 */
export function composeLocationHash(
  coordinates?: Coordinates | undefined,
  streetAddress?: StreetAddress | undefined,
  adminLocation?: AdminLocation | undefined,
): string | undefined {
  if (
    coordinates === undefined &&
    streetAddress === undefined &&
    adminLocation === undefined
  ) {
    return undefined;
  }

  const input =
    (coordinates?.hashId ?? "") +
    "|" +
    (streetAddress?.hashId ?? "") +
    "|" +
    (adminLocation?.hashId ?? "");

  return "v1." + sha256Hex(input);
}

File: public/packages/dotnet/location/core/ComposeLocationHash.cs
// -----------------------------------------------------------------------
// <copyright file="ComposeLocationHash.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Location;

using System.Security.Cryptography;
using DcsvIo.D2.Location.ValueObjects;

/// <summary>
/// Free composition function that joins up to three location value
/// objects (<see cref="Coordinates"/> + <see cref="StreetAddress"/> +
/// <see cref="AdminLocation"/>) into a single content-addressable hash
/// identifier. All-null input returns <c>null</c> (location absent —
/// not an error); any non-null subset returns a deterministic
/// <c>"v1."</c>-prefixed SHA-256 hex digest.
/// </summary>
public static class ComposeLocationHash
{
    /// <summary>
    /// Composes a single hash identifier from the (optional) three
    /// location components. Returns <c>null</c> when ALL three inputs
    /// are null (location is absent — not an error). Otherwise returns
    /// <c>"v1." + SHA-256(c.HashId | s.HashId | a.HashId)</c>; missing
    /// slots contribute <c>""</c> (positional, never collapsed). Inner
    /// component <c>"v1."</c> prefixes ARE included in the outer hash
    /// input.
    /// </summary>
    /// <remarks>
    /// Returns <c>string?</c> rather than <c>D2Result&lt;string&gt;</c>
    /// — documented §17 carve-out: the operation cannot fail (inputs
    /// are already-validated VOs or null), and all-null is a legitimate
    /// non-error state. Wrapping with <c>D2Result</c> would force
    /// consumers to unwrap an always-<c>Ok</c> result on the data path.
    /// </remarks>
    /// <param name="coordinates">Optional <see cref="Coordinates"/> component.</param>
    /// <param name="streetAddress">Optional <see cref="StreetAddress"/> component.</param>
    /// <param name="adminLocation">Optional <see cref="AdminLocation"/> component.</param>
    /// <returns>The composed hash identifier, or <c>null</c> for all-null input.</returns>
    public static string? Compose(
        Coordinates? coordinates,
        StreetAddress? streetAddress,
        AdminLocation? adminLocation)
    {
        if (coordinates is null && streetAddress is null && adminLocation is null)
            return null;

        var input =
            (coordinates?.HashId ?? string.Empty) + "|" +
            (streetAddress?.HashId ?? string.Empty) + "|" +
            (adminLocation?.HashId ?? string.Empty);

        // BCL static one-shot per §15.8 — no IDisposable instance to manage.
        var hashBytes = SHA256.HashData(System.Text.Encoding.UTF8.GetBytes(input));
        return "v1." + Convert.ToHexStringLower(hashBytes);
    }
}
